programs/gateway/src/instructions/update_voter_weight_record.rs

```diff
@@ -56,14 +52,92 @@ pub fn update_voter_weight_record(
 
-    voter_weight_record.voter_weight = DEFAULT_VOTE_WEIGHT;
+    let input_voting_weight_account = ctx.accounts.input_voting_weight.to_account_info();
+
+    let clone_record = voter_weight_record.clone();
+    let input_voter_weight_record = parse_input_voter_weight_record(
+        &input_voting_weight_account,
+        &clone_record,
+        &ctx.accounts.registrar,
+    )?;
 
-    // Record is only valid as of the current slot
-    voter_weight_record.voter_weight_expiry = Some(Clock::get()?.slot);
+    msg!(
+        "input_voter_weight_record.voter_weight: {}",
+        input_voter_weight_record.get_voter_weight()
+    );
+    voter_weight_record.voter_weight = input_voter_weight_record.get_voter_weight();
+    voter_weight_record.weight_action = input_voter_weight_record.get_weight_action();
+    voter_weight_record.weight_action_target = input_voter_weight_record.get_weight_action_target();
 
-    // Set the action to make it specific and prevent being used for voting
-    voter_weight_record.weight_action = Some(voter_weight_action);
-    voter_weight_record.weight_action_target = target;
+    // If the input voter weight record has an expiry, use the max between that and the current slot
+    // Otherwise use the current slot
+    let current_slot = Clock::get()?.slot;
+    voter_weight_record.voter_weight_expiry = input_voter_weight_record.get_vote_expiry().map_or(
+        Some(current_slot), // no previous expiry, use current slot
+        |previous_expiry| Some(max(previous_expiry, current_slot)),
+    );
 
     Ok(())
 }
+
+/// Attempt to parse the input account as a VoterWeightRecord or a TokenOwnerRecordV2
+fn parse_input_voter_weight_record<'a>(
+    input_account: &'a AccountInfo,
+    voter_weight_record_to_update: &'a VoterWeightRecord,
+    registrar: &'a Registrar,
+) -> Result<Box<dyn GenericVoterWeight + 'a>> {
+    let predecessor_generic_voter_weight_record =
+        parse_input_voter_weight_record_unchecked(input_account, registrar)?;
+
+    // ensure that the correct governance token is used
+    require!(
+        voter_weight_record_to_update.governing_token_mint
+            == predecessor_generic_voter_weight_record.get_governing_token_mint(),
+        GatewayError::InvalidPredecessorVoterWeightRecordGovTokenMint
+    );
+
+    // Ensure that the correct governance token is used
+    require!(
+        voter_weight_record_to_update.governing_token_owner
+            == predecessor_generic_voter_weight_record.get_governing_token_owner(),
+        GatewayError::InvalidPredecessorVoterWeightRecordGovTokenOwner
+    );
+
+    // Ensure that the realm matches the current realm
+    require!(
+        registrar.realm == predecessor_generic_voter_weight_record.get_realm(),
+        GatewayError::InvalidPredecessorVoterWeightRecordRealm
+    );
+
+    Ok(predecessor_generic_voter_weight_record)
+}
+
+fn parse_input_voter_weight_record_unchecked<'a>(
```

Comment:
nit: this name is bit confusing because it does check the `owner` for both accounts.
It could be  `get_generic_voter_weight_record_data` to be consistent with `spl-gov` naming convention